mp-mpiu.h

```diff
@@ -0,0 +1,92 @@
+#ifndef _MPIU_H_
+#define _MPIU_H_
+
+typedef void * (*mpiu_malloc_func)(const char * name, size_t size, const char * file, const int line, void * userdata);
+typedef void (*mpiu_free_func)(void * ptr, const char * file, const int line, void * userdata);
+
+void * mpiu_malloc(const char * name, size_t size, const char * file, const int line);
+void mpiu_free(void * ptr, const char * file, const int line);
+
+void mpiu_set_malloc(mpiu_malloc_func malloc, mpiu_free_func free, void * userdata);
+void MPIU_Set_verbose_malloc(MPI_Comm comm);
```

Comment:
I have two concerns about this. The first is: why not just use the MP-Gadget memory allocator, which we already have and which already supports these things? Is it because of the python allocation override?

The second is that I am worried that the function pointer might seriously affect code performance. Not because of the pointer exactly, but because it will stop the compiler from inlining a bunch of allocations if it wants to. Have you checked the effect of this?

I know you don't like macros (I don't like macros) but I feel that this might just be much easier if we use one here. We don't really need to change the memory allocator at runtime, after all.

Reply:
I think you missed the comment that MPIU_Malloc is only supposed to be used for `significant` allocations, e.g. allocations that needs to be freed in a particularly way by the caller, or allocations that needs to be tracked externally. mpiu_set_malloc provides a way to do that.

Base on that, I did not think we shall convert every malloc to mpiu_malloc (also see my above comment), but only those that are particularly large or expected by the caller to be freed in a particular way. Based on that principle, then there really isn't much concern about performance. 'Significant allocations' are by definition slow and infrequent.

Concern 1. This PR does not introduce a new memory allocator. It only introduces a protocol to swap out the allocator for 'significant' allocations. The interface designed here is to allow MP-Gadget to add a function call

mpiu_set_malloc(mymalloc_adaptor, myfree_adaptor, NULL)

in mymalloc_init. Then MPIU will pick up the MP-Gadget allocator, such that the memory it allocates can be freed by myfree, and we do not violate the memory limit.

In addition to Python, FastPM has its own `quirky memory allocator`, and it differs from MP-Gadget's. The interface here is general enough to adapt to all three of them.

Concern 2. For one, the glibc malloc is also implemented as a function pointer: https://www.gnu.org/software/libc/manual/html_node/Hooks-for-Malloc.html
If glibc doesn't care about the pointer redirection, then we probably shouldn't be too concerned.

More importantly though, because I am only using MPIU_Malloc for significant allocations, MPIU_Malloc is only used in very few places in the code current code:

1. allocating the out buffer of MPIU_Gather and MPIU_Scatter. 
2. allocating the temporary buffer on the group leader rank for 'gather-sort'.

Both portion of code are surrounded by expensive MPI calls; there can't be any performance difference due to a pointer.
As we port functions to MPIU from MP-Gadget we will get more users of MPIU_Malloc. but they will likely be used in similar manners, immediately followed by expensive MPI calls.

Reply:
hah, fair enough. If glibc uses a pointer anyway then indeed it is fine!